# Advent of Code 2020 Day 6: Custom Customs

## Import libraries

In [1]:
import numpy as np

## Part 1

### Description Part 1

As your flight approaches the regional airport where you'll switch to a much larger plane, customs declaration forms are distributed to the passengers.

The form asks a series of 26 yes-or-no questions marked a through z. All you need to do is identify the questions for which anyone in your group answers "yes". Since your group is just you, this doesn't take very long.

However, the person sitting next to you seems to be experiencing a language barrier and asks if you can help. For each of the people in their group, you write down the questions for which they answer "yes", one per line. For example:

abcx
abcy
abcz
In this group, there are 6 questions to which anyone answered "yes": a, b, c, x, y, and z. (Duplicate answers to the same question don't count extra; each question counts at most once.)

Another group asks for your help, then another, and eventually you've collected answers from every group on the plane (your puzzle input). Each group's answers are separated by a blank line, and within each group, each person's answers are on a single line. For example:

abc

a
b
c

ab
ac

a
a
a
a

b
This list represents answers from five groups:

The first group contains one person who answered "yes" to 3 questions: a, b, and c.
The second group contains three people; combined, they answered "yes" to 3 questions: a, b, and c.
The third group contains two people; combined, they answered "yes" to 3 questions: a, b, and c.
The fourth group contains four people; combined, they answered "yes" to only 1 question, a.
The last group contains one person who answered "yes" to only 1 question, b.
In this example, the sum of these counts is 3 + 3 + 3 + 1 + 1 = 11.

For each group, count the number of questions to which anyone answered "yes". What is the sum of those counts?

### Import data

In [2]:
with open('day_6_input.txt') as input:
    forms = [line.strip() for line in input.read().split('\n\n')]

In [3]:
forms[:10]

['vkplsqwiftuazyje\nmokluxwbsfhgc',
 'tlfxgqinzmdju\nzjmvfstnplqd\nxztfjnlqemd\ncmatrlfnbjqodz',
 'xrumszojqa\nmqsorcixzuja',
 'bwdvujyzsneiom\nnjqzvyhceriodl',
 'nhbxcfgwpraqje\nabcqfnrpwehgxj\npqrfxancbjglewh\nexhfbanpwrcgjq',
 'skmy\nkoq\nzqtk\nvkgfjun',
 'zkmhpcq\njnubrdys\nfvdswier',
 'nj\nnjehcry\nyhnr\ntudngxkovfa',
 'pcnv\nnscvp\nnvpc',
 'fjmpugzckbwho\nquxmhkzad\nuzmohyjkr']

In [4]:
# Remove the separations between responses for each group
forms_no_sep = [item.replace('\n', '') for item in forms]

In [5]:
forms_no_sep[:10]

['vkplsqwiftuazyjemokluxwbsfhgc',
 'tlfxgqinzmdjuzjmvfstnplqdxztfjnlqemdcmatrlfnbjqodz',
 'xrumszojqamqsorcixzuja',
 'bwdvujyzsneiomnjqzvyhceriodl',
 'nhbxcfgwpraqjeabcqfnrpwehgxjpqrfxancbjglewhexhfbanpwrcgjq',
 'skmykoqzqtkvkgfjun',
 'zkmhpcqjnubrdysfvdswier',
 'njnjehcryyhnrtudngxkovfa',
 'pcnvnscvpnvpc',
 'fjmpugzckbwhoquxmhkzaduzmohyjkr']

In [6]:
# Create a list of sublists to test the function
test_list = [['abc'], ['a', 'b', 'c'], ['ab', 'ac'], ['a', 'a', 'a', 'a'], ['b']]

# Create a list with no sublists to test the function
test_no_sep = ['abc', 'abc', 'abac', 'aaaa', 'b']

# Create a dictionary of answers to check
testdic = {'abc': 3, 'a,b,c': 0, 'ab,ac': 1, 'a,a,a,a': 1, 'b':1}

print(test_list, test_no_sep, testdic)

[['abc'], ['a', 'b', 'c'], ['ab', 'ac'], ['a', 'a', 'a', 'a'], ['b']] ['abc', 'abc', 'abac', 'aaaa', 'b'] {'abc': 3, 'a,b,c': 0, 'ab,ac': 1, 'a,a,a,a': 1, 'b': 1}


In [7]:
# Create a list of sublists to store the form responses

# Make an empty list to store the sublists
forms_list = []
for entries in forms:
    # Make an empty list to store each sublist as it's made
    current_list = []
    # Split the entries (to make into a sublist) by the delimiter
    add = entries.split('\n')
    # Add split entries to sublist
    current_list += add
    # Add sublist to list of sublists
    forms_list.append(current_list)

In [8]:
forms_list[:10]

[['vkplsqwiftuazyje', 'mokluxwbsfhgc'],
 ['tlfxgqinzmdju', 'zjmvfstnplqd', 'xztfjnlqemd', 'cmatrlfnbjqodz'],
 ['xrumszojqa', 'mqsorcixzuja'],
 ['bwdvujyzsneiom', 'njqzvyhceriodl'],
 ['nhbxcfgwpraqje', 'abcqfnrpwehgxj', 'pqrfxancbjglewh', 'exhfbanpwrcgjq'],
 ['skmy', 'koq', 'zqtk', 'vkgfjun'],
 ['zkmhpcq', 'jnubrdys', 'fvdswier'],
 ['nj', 'njehcry', 'yhnr', 'tudngxkovfa'],
 ['pcnv', 'nscvp', 'nvpc'],
 ['fjmpugzckbwho', 'quxmhkzad', 'uzmohyjkr']]

### Find number of unique characters in each string

In [9]:
# Create an empty list to store counts for each group
group_count = []
for i in forms_no_sep:
    # Use set() to remove duplicate characters from the entry
    i = set(i)
    # Get the length of the entry (if all unique, = number of unique)
    count = len(i)
    # Add each group count to the list of group counts
    group_count += [count]

In [10]:
# Convert to a numpy array for math
group_count = np.array(group_count)

In [11]:
# Calculate the total sum of 'yes' answers
total_sum = np.sum(group_count)

print(total_sum)

6382


## Solution: The sum of questions with an answer of 'yes', unique to each group, is 6382.

## Part 2

### Description Part 2

As you finish the last group's customs declaration, you notice that you misread one word in the instructions:

You don't need to identify the questions to which anyone answered "yes"; you need to identify the questions to which everyone answered "yes"!

Using the same example as above:

abc

a
b
c

ab
ac

a
a
a
a

b
This list represents answers from five groups:

In the first group, everyone (all 1 person) answered "yes" to 3 questions: a, b, and c.
In the second group, there is no question to which everyone answered "yes".
In the third group, everyone answered yes to only 1 question, a. Since some people did not answer "yes" to b or c, they don't count.
In the fourth group, everyone answered yes to only 1 question, a.
In the fifth group, everyone (all 1 person) answered "yes" to 1 question, b.
In this example, the sum of these counts is 3 + 0 + 1 + 1 + 1 = 6.

For each group, count the number of questions to which everyone answered "yes". What is the sum of those counts?

In [12]:
# Compare within sublists & add 1 to total for group if all entries in sublist have the same letter(s)

def agreed_yes_count(input):
    # Create an empty list to store the count of agreed yes per group
    groups_unique = []
    # Set input index to zero to start loop
    input_index = 0
    # Loop through the larger list
    while input_index < len(input):
        # Loop through each sublist in list
        sublist = input[input_index]
        # Reset the character index to zero for next sublist
        response_index = 0
        # Create an empty list to store the unique characters for the entire group
        group_dic = {}
        # Reset group count to zero for next sublist
        group_count = 0
        # Loop through the sublists
        for responses in sublist:
            # loop through each response in sublist
            response = sublist[response_index]
            # Reset character index
            char_index = 0
            # Loop through the characters
            for char in responses:
                # Loop through each character in response
                char = responses[char_index]
                # Reset value back to zero for each loop
                value = 1
                # Create a new key if character isn't already a key in dictionary
                if char not in group_dic:
                    # Set the key equal to a value of 1 if not already in dictionary
                    group_dic[char] = value
                else:
                    # Add 1 to value if character already in dictionary
                    group_dic[char] += 1
                # Move to the next character
                char_index += 1
            # Define what the count needs to be for all individuals to have responded yes
            target_count = len(sublist)
            # Iterate through dictionary to see if number of yes match the number of responses
            for key, value in group_dic.items():
                # Check if the number of yes responses for that characcter is equal to the number of responders
                if value == target_count:
                    # If it is equal, increase the count for the group by one to indicate a unanimous yes
                    group_count += 1
            # Move to the next response
            response_index += 1
        # Add the character count for the individual group to the larger list for all groups
        groups_unique.append(group_count)
        # Move to the next sublist
        input_index += 1
    # Convert list to numpy array
    groups_unique = np.array(groups_unique)
    
    return groups_unique

In [13]:
# Test that the function is working properly on the test set
test_array = agreed_yes_count(test_list)

print(test_array)

[3 0 1 1 1]


In [14]:
# Check against test set correct answers
testdic

{'abc': 3, 'a,b,c': 0, 'ab,ac': 1, 'a,a,a,a': 1, 'b': 1}

In [15]:
# Check test sum is correct
test_sum = sum(test_array)

print(test_sum)

6


In [16]:
# Run the function on the input
groups_agreed_yes_count = agreed_yes_count(forms_list)

# Calculate the total sum of unanimous yes reponses
total_sum = sum(groups_agreed_yes_count)

print(total_sum)

3197


## Solution: The number of unanimous 'yes' responses for the groups is 3197.